go/chat/utils/utils.go

```diff
@@ -190,6 +190,55 @@ func splitAndNormalizeTLFNameCanonicalize(mctx libkb.MetaContext, name string, p
 	return writerNames, readerNames, extensionSuffix, err
 }
 
+// AttachContactNames retrieves display names for SBS phones/emails that are in the phonebook.
+func AttachContactNames(mctx libkb.MetaContext, participants []chat1.ConversationLocalParticipant) (withContacts []chat1.ConversationLocalParticipant) {
+	var contacts []keybase1.ProcessedContact
+	var err error
+	contactsFetched := false
+	for _, participant := range participants {
+		if isPhoneOrEmail(participant.Username) {
+			if !contactsFetched {
+				contacts, err = mctx.G().SyncedContactList.RetrieveContacts(mctx)
+				if err != nil {
+					mctx.Debug("Error fetching contacts: %s", err)
+					return participants
+				}
+			}
+			// todo separate phone / email from assertion
+			assertion, err := libkb.ParseAssertionURL(mctx.G().MakeAssertionContext(mctx), participant.Username, true)
+			if err == nil {
+				phoneOrEmail := assertion.GetValue()
+				isPhone := assertion.GetKey() == "phone"
+				contactName := findContactName(contacts, phoneOrEmail, isPhone)
+				participant.ContactName = contactName
+			} else {
+				mctx.Debug("Error parsing assertion: %s", err)
+			}
+			withContacts = append(withContacts, participant)
+		}
+	}
+	return withContacts
+}
+
+var nonDigits = regexp.MustCompile("[^\\d]")
+
+func findContactName(contacts []keybase1.ProcessedContact, phoneOrEmail string, isPhone bool) *string {
+	for _, contact := range contacts {
+		cPhoneOrEmail := contact.Component.ValueString()
+		if isPhone {
+			cPhoneOrEmail = nonDigits.ReplaceAllString(cPhoneOrEmail, "")
+		}
+		if cPhoneOrEmail == phoneOrEmail {
+			return &contact.ContactName
+		}
+	}
+	return nil
+}
+
+func isPhoneOrEmail(username string) bool {
```

Comment:
@zapu is there a better way to do this?

Reply:
I won't I see, this is kind of a heuristic?

Reply:
it would be better to use assertion parser and look at assertion key, but this would work too, not as elegantly though.

Reply:
This is used in `PresentConversationLocal` which can't create an `AssertionContext` for the parser. The use in `AttachContactNames` could use the parser, though I thought this was a nice short circuit so we wouldn't run it on every participant in the inbox.